---
title: "R Notebook"
output: html_notebook
---

### Load libraries
```{r}
library("AnnotationDbi")
library("edgeR")
library(GenomicAlignments)
library(ggplot2)
library(ggrepel)
library(heatmap3)
library("Homo.sapiens")
library(pcaMethods)
library(SummarizedExperiment)
library(tidyverse)
```

### Load the data and join the raw count files into a dataframme
```{r}
## Gene expression data
# List files in the folder
files = list.files('~/GitHubrepo/cancer_analyses_python/PhD_code/')

# Read the first file
se <- readRDS(paste('~/GitHubrepo/cancer_analyses_python/PhD_code/',files[1], sep=""), refhook = NULL)
countdata <- as.data.frame(assay(se)) %>% 
      rownames_to_column(., 'gene_id')

# Join the remaining files into countdata
for (x in files) { 
  if (x != files[1]){
    se <- readRDS(paste('~/GitHubrepo/cancer_analyses_python/PhD_code/', x, sep=""), refhook = NULL)
    countdata_sample <- as.data.frame(assay(se)) %>% 
      rownames_to_column(., 'gene_id')
    countdata <- merge(countdata, countdata_sample, by = 'gene_id')
  }
}

# Format
countdata = countdata %>% rename_with(~ str_remove(., ".bam"), everything())


## Sample data
sample_data <- data.frame(row.names  = c("ASOM2", "ASOM4", "ASOM5", "NCM2", "NCM4", "NCM6"),
                           dex = factor(c("trt", "trt", "trt", "untrt", "untrt", "untrt"), levels = c("untrt", 'trt')))

sample_data$dex #Check that untreated or NC is the first level, the one that will be taken as reference. 
```


### Normalise counts with edgeR
```{r}
# Create a DGEList object, group indicator for each column
y <- DGEList(counts=as.matrix(countdata[, -1]), 
             samples=sample_data , 
             genes=countdata$gene_id,
             group = factor(rep(c("trt", "untrt"), each=3)))

# Determine which genes have sufficiently large counts to be retained in a statistical analysis.
# The filterByExpr function keeps rows that have worthwhile counts in a min N of samples 
keep <- filterByExpr(y)
table(keep) #FALSE = 45428, TRUE =18249

# Remove lowly expressed genes
y <- y[keep, ,keep.lib.sizes=FALSE]

# Calculate normalization factors to scale the raw library sizes
dge <- calcNormFactors(y)

# Get TMM normalised counts
tmm <- cpm(dge, log = TRUE)
head(tmm)
```



### QC
```{r}
#Compute PCA (data is mean centered but not scaled as it's already normalised ) 
md  <- prep(tmm, scale="none", center=TRUE) #centre the data
pc <- pca(t(md), method="svd")


# We don't do crossvalidaiton, because svd method doesn't allow it. Otherwise would be something like this: q2PCA <- Q2(resPCA, md, fold=10)
pc_scores <- scores(pc)
pc.scores <- as.data.frame(pc_scores) #Convert it into a dataframe
pc.scores$Treatment <- (c("ASO361", "ASO361","ASO361","NC", "NC", "NC" )) #This cannot happen unless it is a data frame

#Format
rownames(pc.scores) <- gsub("[.bam]", "", rownames(pc.scores)) 

#Get a short summary on the calculated mode
#summary(resPCA)

pca_plot <- ggplot(pc.scores, aes(PC1, PC2), color=ttment) +
  geom_point(aes(color = Treatment)) +
  #geom_text_repel(aes(label=row.names(pc.scores))) +
  xlab(paste("PC1", pc@R2[1] * 100, "% of the variance")) +
  ylab(paste("PC2", pc@R2[2] * 100, "% of the variance")) +
  ggtitle("PCA plot") +
  theme(plot.title = element_text(hjust = 0.5, face="bold") )
pca_plot
```

### Perform differential gene expression
```{r}
# Prepare the design matrix: comparison contrast, data.
design <- model.matrix(~ dex, data = dge$samples)

# Estimate common dispersion and tagwise dispersions in one run
dge <- estimateDisp(dge, design)

# Perform likelihood ratio tests
# Note that this approach will only be successful if the number of DE genes is relatively small
fit <- glmFit(dge, design)
lrt <- glmLRT(fit, coef=ncol(design))

# Extracts the most DEG from a test object, ranked either by p-value.
tt <- topTags(lrt, n=nrow(dge))

# Double check the comaprison - prints "dextrt", which is correct
tt$comparison 

# Get df containing differential expression results for the top genes in sorted order
tt_table <- tt$table

# Find how many siignificant DEG were identified
table(tt_table$FDR<0.05) #FALSE:18118,  TRUE 131

# Identify which genes are significantly DE 
summary(decideTests(lrt, adjust.method = "fdr")) 
#        dextrt
#Down       90
#NotSig  18118
#Up         41

summary(decideTests(lrt, adjust.method = "fdr", lfc = 1)) #Putting the logFC threshold.

#         dextrt
#Down       7
#NotSig  18214
#Up         28

# Get the top 10 DEG 
tt10 <- topTags(lrt)
tt10
```


### Annotate ensembl IDs with gene symbols
```{r}
tt_table$symbol <- mapIds(Homo.sapiens,
                          keys=tt_table$genes,
                          column="SYMBOL",
                          keytype="ENSEMBL",
                          multiVals="first")
```


### Volcano plot with significant DEG
```{r}
# Make a copy of the data
data<- tt_table
data <- as.data.frame(data)

# Add a column denoting if gene is significantly DE
data$Significant <- ifelse(data$FDR < 0.05, "FDR < 0.05", "Not Sig")

# Plot volcano plot
ggplot(data, aes(x = logFC, y = -log10(FDR))) +
geom_point(aes(color = Significant)) +
  scale_color_manual(values = c("red", "grey")) +
  theme_bw(base_size = 12) + theme(legend.position = "bottom") +
  geom_text_repel(
    data = subset(data, FDR < 0.01),
    aes(label = symbol),
    size = 3
  )
```